File: Matlab-code/runValidation.m
%run validation

dataPath = './';
modelPath = './';
savePath = './';
nameSavedModel = 'trainedModel_N1000';

%load saved trained model
load([modelPath,nameSavedModel])


W = double(model.W);
Vall = model.Vall;
betaAll = model.betaAll;
trainBasisFunMean = model.trainBasisFunMean;
trainBasisFunStd = model.trainBasisFunStd;
imTrainMean = model.imTrainMean;
imTrainStd = model.imTrainStd;
indecesMask = model.indecesMask;
nLatentValues = model.nLatentValues;
nLatentTot = numel(nLatentValues);


%load validation data 

load([dataPath,'age_uk_healthy_valid.mat']) %load age_healthy_valid
%array of size (# of subjects, 1) with age of all validation subjects

targetValid=age_healthy_valid;
nValid=size(targetValid,1);
covariatesValid=[];
clear('age_healthy_valid')


%validation basis functions

validBasisFun=[targetValid,covariatesValid];
stValidBasisFun=(validBasisFun-trainBasisFunMean)./trainBasisFunStd;


%validation images

load([dataPath,'T1_nonlin_uk_healthy_valid_down3.mat']) %load T1_nonlin_down3
%load 4D array of size (# of subjects, image dimension 1, image dimension 2, image dimension 3)
% with images of all validation subjects

allVolumes = T1_nonlin_down3;
clear T1_nonlin_down3

imValid = zeros(nValid,numel(indecesMask));
for nSubj = 1:nValid
    vol = squeeze(allVolumes(nSubj,:,:,:));
    imValid(nSubj,:) = vol(indecesMask);
end
clear('allVolumes')


stValidImages = (imValid-imTrainMean)./imTrainStd;
clear('imValid')




%initialization
correlations=zeros(nLatentTot,1);
MAEs=zeros(nLatentTot,1);
RMSEs=zeros(nLatentTot,1);
predictions=cell(nLatentTot,1);


% target predictions

   
   for n = 1:nLatentTot

    nLatentVars=nLatentValues(n);
    fprintf('n.latent = %d\n',nLatentVars)
    V = double(Vall{n,1});
    beta = betaAll{n,1};
  
   
    [targetPredictions,correlation,MAE,RMSE] = testModel(stValidImages,stValidBasisFun,targetValid,V,beta,W,trainBasisFunMean,trainBasisFunStd);

  

    %compute correlation and error

    correlations(n)=correlation;
    RMSEs(n)=RMSE;
    MAEs(n)=MAE;
    predictions{n}=single(targetPredictions);

   end




   validResults.nLatentValues=nLatentValues;
   validResults.correlations=correlations;
   validResults.MAEs= MAEs;
   validResults.RMSEs= RMSEs;
   validResults.predictions=predictions;


   save([savePath,'validResults'],'validResults','-v7.3')

   

File: runValidation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""

"""

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""

"""

import scipy as sp
import numpy as np
import hdf5storage
import pickle
import argparse

from testModel import *


parser = argparse.ArgumentParser(description="Validate the model",
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)


parser.add_argument("-mp","--modelPath", help="path where the trained model is saved", default="./")
parser.add_argument("-n","--nameSavedModel", help="name of the saved model",default="trainedModel.pkl")
parser.add_argument("-dp","--dataPath", help="path to validation data", default="./")



args = parser.parse_args()
config = vars(args)
print(config)

modelPath = config['modelPath']
nameSavedModel = config['nameSavedModel']
dataPath = config['dataPath']



#load trained model

modelPklFile = modelPath + nameSavedModel   
with open(modelPklFile, 'rb') as file:  
    model = pickle.load(file)
    
W = model["W"]
Vall = model["Vall"] 
betaAll = model["betaAll"]
trainBasisFunMean = model["trainBasisFunMean"]
trainBasisFunStd = model["trainBasisFunStd"]
imTrainMean = model["imTrainMean"]
imTrainStd = model["imTrainStd"]
indecesMask = model["indecesMask"]
nLatentValues = model["nLatentValues"]


#load validation data

age = np.load(dataPath + 'validAge.npy')
# array of size (# of subjects, 1) with age of all validation subjects
allVolumes = np.load( dataPath + 'validImages.npy')
# 4D array of size (# of subjects, image dimension 1, image dimension 2, image dimension 3)
# with images of all validation subjects


validTarget = age
Nvalid = len(validTarget)
covariatesValid = []


   
#  test basis functions
 
validBasisFun = validTarget
stValidBasisFun=(validBasisFun-trainBasisFunMean)/trainBasisFunStd
 
 


imValid=np.zeros((Nvalid,np.sum(indecesMask)))
for n in np.arange(Nvalid):
    vol = np.squeeze(allVolumes[n,:,:,:])
    imValid[n,:] = np.transpose(vol[indecesMask])
del allVolumes 

stValidImages=(imValid-imTrainMean)/imTrainStd 


#validation of model

nLatentTot = len(nLatentValues)
correlations = np.zeros([nLatentTot,1])
MAEs = np.zeros([nLatentTot,1])
RMSEs = np.zeros([nLatentTot,1])
targetPredictionsAll = list()

n = 0
for nLatentVars in nLatentValues:
    
    print("N. latent = " + str(nLatentVars) + "\n")
    V = Vall[n]
    beta = betaAll[n]
    
    (targetPredictions,correlation,MAE,RMSE) = testModel(stValidImages,stValidBasisFun,validTarget,V,beta,W,trainBasisFunMean,trainBasisFunStd)
    
    correlations[n] = correlation
    MAEs[n] = MAE
    RMSEs[n] = RMSE 
    targetPredictionsAll.append(targetPredictions)
    
    n = n+1
    
    
    
validResults = {
  "targetPredictionsAll": targetPredictionsAll,
  "correlations": correlations,
  "MAEs": MAEs,
  "RMSEs": RMSEs
}

#save test results

validResultsPath = './'
validPklFile = validResultsPath + "validationResults.pkl"  

with open(validPklFile, 'wb') as file:  
    pickle.dump(validResults, file)

